Source/Core/Core/Slippi/SlippiPlayback.cpp

```diff
@@ -0,0 +1,272 @@
+	cv_processingDiff.notify_one();
+	std::string diff = std::string();
+	open_vcdiff::VCDiffEncoder encoder((char *)iState.data(), iState.size());
+	encoder.Encode((char *)cState.data(), cState.size(), &diff);
+
+	INFO_LOG(SLIPPI, "done processing");
+	numDiffsProcessing -= 1;
+	cv_processingDiff.notify_one();
+	return diff;
+}
+
+SlippiPlaybackStatus::SlippiPlaybackStatus()
+{
+	shouldJumpBack = false;
+	shouldJumpForward = false;
+	inSlippiPlayback = false;
+	shouldRunThreads = false;
+	isHardFFW = false;
+	isSoftFFW = false;
+	lastFFWFrame = INT_MIN;
+	currentPlaybackFrame = INT_MIN;
+	targetFrameNum = INT_MAX;
+	latestFrame = Slippi::GAME_FIRST_FRAME;
+}
+
+void SlippiPlaybackStatus::startThreads()
+{
+	shouldRunThreads = true;
+	m_savestateThread = std::thread(&SlippiPlaybackStatus::SavestateThread, this);
+	m_seekThread = std::thread(&SlippiPlaybackStatus::SeekThread, this);
+}
+
+void SlippiPlaybackStatus::prepareSlippiPlayback()
+{
+	// block if there's too many diffs being processed
+	while (numDiffsProcessing > 3)
+	{
+		INFO_LOG(SLIPPI, "Processing too many diffs, blocking main process");
+		cv_processingDiff.wait(processingLock);
+	}
+
+	// Unblock thread to save a state every interval
+	if ((currentPlaybackFrame + 122) % FRAME_INTERVAL == 0)
+		condVar.notify_one();
+
+	// TODO: figure out why sometimes playback frame increments past targetFrameNum
+	if (inSlippiPlayback && currentPlaybackFrame >= targetFrameNum)
+	{
+		if (currentPlaybackFrame > targetFrameNum)
+		{
+			INFO_LOG(SLIPPI, "Reached frame %d. Target was %d. Unblocking", currentPlaybackFrame,
+			         targetFrameNum);
+		}
+		cv_waitingForTargetFrame.notify_one();
+	}
+}
+
+void SlippiPlaybackStatus::resetPlayback()
+{
+	shouldRunThreads = false;
+
+	if (m_savestateThread.joinable())
+		m_savestateThread.detach();
+
+	if (m_seekThread.joinable())
+		m_seekThread.detach();
+
+	condVar.notify_one(); // Will allow thread to kill itself
+
+	shouldJumpBack = false;
+	shouldJumpForward = false;
+	targetFrameNum = INT_MAX;
+	inSlippiPlayback = false;
+	futureDiffs.clear();
+	futureDiffs.rehash(0);
+}
+
+void SlippiPlaybackStatus::processInitialState(std::vector<u8> &iState)
+{
+	INFO_LOG(SLIPPI, "saving iState");
+	State::SaveToBuffer(iState);
+	SConfig::GetInstance().bHideCursor = false;
+};
+
+void SlippiPlaybackStatus::SavestateThread()
+{
+	Common::SetCurrentThreadName("Savestate thread");
+	std::unique_lock<std::mutex> intervalLock(mtx);
+
+	INFO_LOG(SLIPPI, "Entering savestate thread");
+
+	while (shouldRunThreads)
+	{
+		// Wait to hit one of the intervals
+		// Possible while rewinding that we hit this wait again.
+		while (shouldRunThreads && (currentPlaybackFrame - Slippi::PLAYBACK_FIRST_SAVE) % FRAME_INTERVAL != 0)
+			condVar.wait(intervalLock);
+
+		if (!shouldRunThreads)
+			break;
+
+		s32 fixedFrameNumber = currentPlaybackFrame;
+		if (fixedFrameNumber == INT_MAX)
+			continue;
+
+		bool isStartFrame = fixedFrameNumber == Slippi::PLAYBACK_FIRST_SAVE;
```

Comment:
change the start frame to -122 instead of -123. this saves a whole lot of loading when dmawrite reloads a ton of assets because it expects to be starting a game from scratch, when really this is just a save state so its kind of pointless to be doing that. Also prevents other crashes where calling DMAWrite and loading assets for initial -123 frame over and over breaks the main rendering loop